### PR TITLE
NOT QA READY: OSVR HDK support

### PR DESCRIPTION
**Pull request by ctrlaltdavid**

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: f7a95695bd2d71c680f663daf5cb83e69d5bf1ca available at: https://gist.github.com/5bf597952086b96e14504d60199c52be<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-f7a9569.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-f7a9569.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9b09f2336fad424c9bc6633b11a9da32a17a3794 available at: https://gist.github.com/5bf597952086b96e14504d60199c52be<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-9b09f23.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-9b09f23.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 79d7e5ab98d6a61859912a747587bd0f8a45c199 available at: https://gist.github.com/5bf597952086b96e14504d60199c52be<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-79d7e5a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-79d7e5a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 90134d8bc841acffa47f40bd1d0df16455bc41e9 available at: https://gist.github.com/5bf597952086b96e14504d60199c52be<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-90134d8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-90134d8.dmg

**Comment by thoys** [comment]
Do we know anyone with a OSVR HMD ?

File: plugins/hifiOSVR/src/OsvrDisplayPlugin.cpp [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        return false;
+    }
+    _osvrContext->update();
+
+    auto hmdInfo = parseHMDInfo(_osvrContext->getStringParameter("/display"));
+    qDebug() << "OSVR: HMD =" << hmdInfo[HMD_VENDOR] << hmdInfo[HMD_MODEL] << hmdInfo[HMD_VERSION];
+
+    // FIXME: Interface crashes at the following line if start up in HMD mode; OSVR SDK 0.6.1337 build 329.
+    // Work-around implemented as FIXME in Application::loadSettings().
+    osvr::renderkit::RenderManager::OpenResults result = _osvrRender->OpenDisplay();
```
Comment (jherico) [inline plugins/hifiOSVR/src/OsvrDisplayPlugin.cpp]: Rendering stuff may need to be moved to the rendering thread.  `internalActivate` runs on the main thread, while stuff related to OpenGL should happen in `customizeContext`, which happens on the rendering thread.

File: plugins/hifiOSVR/src/OsvrDisplayPlugin.cpp [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        qWarning() << "OSVR: Display does not have 2 eyes";
+        delete _osvrRender;
+        _osvrRender = nullptr;
+        return false;
+    }
+
+    // Get HMD's target frame rate.
+    osvr::renderkit::RenderTimingInfo timingInfo;
+    if (_osvrRender->GetTimingInfo(0, timingInfo)) {
+        const float MICROSECONDS_PER_SECOND = 1000000.0f;
```
Comment (jherico) [inline plugins/hifiOSVR/src/OsvrDisplayPlugin.cpp]: There's an existing USECS_PER_SECOND constant for this.

File: plugins/hifiOSVR/src/OsvrDisplayPlugin.cpp [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    flipProjection(_presentInfo);
+    const bool FLIP_IN_Y = true;
+    if (!_osvrRender->PresentRenderBuffers(_colorBuffers, _presentInfo, _renderParams, _textureViewports, FLIP_IN_Y)) {
+        qWarning() << "OSVR: Failed to present image on HMD";
+    }
+
+    _presentRate.increment();
+};
+
+void OsvrDisplayPlugin::uncustomizeContext() {
```
Comment (jherico) [inline plugins/hifiOSVR/src/OsvrDisplayPlugin.cpp]: You should also reset the `_outputFramebuffer` here

File: plugins/hifiOSVR/src/OsvrHelpers.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    // Interface needs this in order to get the correct projection values (though this could be fixed in goGlm().
+    // OSVR RenderManager also needs this to be done when flipping y in PresentRenderBuffers(), otherwise time-warp is jerky.
+    std::swap(renderInfo[0].projection.top, renderInfo[0].projection.bottom);
+    std::swap(renderInfo[1].projection.top, renderInfo[1].projection.bottom);
+}
+
+inline glm::mat4 toGlm(const osvr::renderkit::OSVR_ProjectionMatrix &projection) {
+    GLdouble proj[16];
+    OSVR_Projection_to_OpenGL(proj, projection);
+    return glm::mat4(
```
Comment (jherico) [inline plugins/hifiOSVR/src/OsvrHelpers.h]: `glm::make_mat4(proj);`

**Comment by jherico** [comment]
@thoys I have one.

**Comment by hifi-gustavo** [comment]
Build console output for commit: d1f423cd2f06fb1f815931ea7f1c70b7d9e5cdcc available at: https://gist.github.com/5bf597952086b96e14504d60199c52be<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-d1f423c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-d1f423c.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 2d7a8478d2e6be676fd5efbbd9fe46af2ea69c67 available at: https://gist.github.com/5bf597952086b96e14504d60199c52be<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-2d7a847.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-2d7a847.dmg

**Comment by ChristophHaag** [comment]
I can only give few comments for Linux.

The whole content of CMakeLists.txt is inside a `if (WIN32)`, so I started by removing that if.

```
    add_dependency_external_projects(boost)
    find_package(boost REQUIRED)
```
did not work for me. Is there some setup that needs to be done for this?
The likewise osvr block didn't work either, but in contrast to this one it didn't throw an error.
It failed to link to the osvr libraries though, so I had to add `target_link_libraries(${TARGET_NAME} osvrClientKit osvrRenderManager)`.

Then the plugin loads, but it makes all the buttons in the toolbar disappear except the HUD toggle button.
btns
Without the osvr plugin the toolbar works normally. May be my fault for compiling with qt 5.7 (because I haven't figured out yet how to compile with non-system installed qt libraries).

But there is still the menu option Display - OSVR HMD to enable it.
This does not work on mesa.
```
Mesa: User error: GL_INVALID_OPERATION in glBindTexture(non-gen name)
Nov 10 00:47:43.440 warning [RenderManager]: RenderManagerOpenGL::PresentEye after texture bind: OpenGL error GL_INVALID_OPERATION(1282)
Nov 10 00:47:43.440 error [RenderManager]: RenderManager::PresentRenderBuffers(): PresentEye failed.
```
That's somewhere deep in rendermanager after plugins/hifiOSVR/src/OsvrDisplayPlugin.cpp:278 is being executed:
```
#0  0x00007fffcb1f550a in _mesa_error (ctx=0x45cdbd0, error=1282, fmtString=0x7fffcb9b750f "glBindTexture(non-gen name)") at main/errors.c:274
#1  0x00007fffcb32f806 in _mesa_BindTexture (target=3553, texName=35) at main/texobj.c:1711
#2  0x00007fff0950d9b9 in osvr::renderkit::RenderManagerOpenGL::PresentEye(osvr::renderkit::RenderManager::PresentEyeParameters) () at /usr/lib/libosvrRenderManager.so
#3  0x00007fff094e8f9c in osvr::renderkit::RenderManager::PresentRenderBuffersInternal(std::vector<osvr::renderkit::RenderBuffer, std::allocator<osvr::renderkit::RenderBuffer> > const&, std::vector<osvr::renderkit::RenderInfo, std::allocator<osvr::renderkit::RenderInfo> > const&, osvr::renderkit::RenderManager::RenderParams const&, std::vector<osvr::renderkit::OSVR_ViewportDescription, std::allocator<osvr::renderkit::OSVR_ViewportDescription> > const&, bool) () at /usr/lib/libosvrRenderManager.so
#4  0x00007fff094e6c5e in osvr::renderkit::RenderManager::PresentRenderBuffers(std::vector<osvr::renderkit::RenderBuffer, std::allocator<osvr::renderkit::RenderBuffer> > const&, std::vector<osvr::renderkit::RenderInfo, std::allocator<osvr::renderkit::RenderInfo> > const&, osvr::renderkit::RenderManager::RenderParams const&, std::vector<osvr::renderkit::OSVR_ViewportDescription, std::allocator<osvr::renderkit::OSVR_ViewportDescription> > const&, bool) () at /usr/lib/libosvrRenderManager.so
#5  0x00007fff09b02a63 in OsvrDisplayPlugin::hmdPresent() (this=0x1c6e190) at /home/chris/oldhome/build/hifi/hifi/plugins/hifiOSVR/src/OsvrDisplayPlugin.cpp:278
#6  0x0000000000e1248b in HmdDisplayPlugin::internalPresent() (this=0x1c6e190)
    at /home/chris/oldhome/build/hifi/hifi/libraries/display-plugins/src/display-plugins/hmd/HmdDisplayPlugin.cpp:189
#7  0x0000000000e1b4ff in OpenGLDisplayPlugin::present() (this=this@entry=0x1c6e190)
    at /home/chris/oldhome/build/hifi/hifi/libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.cpp:631
#8  0x0000000000e207d1 in PresentThread::run() (this=0x31fd8f0) at /home/chris/oldhome/build/hifi/hifi/libraries/display-plugins/src/display-plugins/OpenGLDisplayPlugin.cpp:207
#9  0x00007fffebca6d78 in  () at /usr/lib/libQt5Core.so.5
#10 0x00007ffff62b0454 in start_thread () at /usr/lib/libpthread.so.0
#11 0x00007fffeb0a17df in clone () at /usr/lib/libc.so.6
```

That's where this stuff goes a bit over my head. Chances are that on the proprietary nvidia/amdpgu-pro driver it would work and is only broken on mesa.

**Comment by ctrlaltdavid** [comment]
@ChristophHaag 
This PR adds OSVR HMD support in Windows only. (OSX and/or Linux probably best done separately after merged, sometime.)
The toolbar in the screen snap probably indicates that you don't have any scripts running. (Toolbar seems to be a defined or minimum width - same thing will happen with non-PR build.)

**Comment by ChristophHaag** [comment]
Oh, when it is intentionally windows-only then yes, you can ignore my comment. I just don't really see why. I think it only needs small adjustments how the boost and osvr libraries are handled and then it builds and loads fine on linux. If it needs to be tested, I'm always happy to. :)

In any way, I was wrong to attribute the vanishing buttons to the osvr plugin. It now happens even after a fresh rebuild without the plugin and after deleting the config. There is big trouble with the scripts dialog with my qt version (that is not supported by high fidelity), so I blame it on that, was probably just a happy accident that it actually worked. Sorry for the noise then, I'm just really looking forward to see this working.

**Comment by ctrlaltdavid** [comment]
@ChristophHaag It's only intentionally Windows-only at present because I don't have Linux. Someone should be able to tweak it relatively easily to work in Linux after this PR is merged.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5ab6332ab0f6ac4ac317a170e878d615402878a0 available at: https://gist.github.com/5bf597952086b96e14504d60199c52be<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-5ab6332.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR8628-5ab6332.dmg

**Comment by cozza13** [comment]
We are not going to deploy this at the moment. Contact me for any questions.

**Pull request closed without merging**